website/blog/2022-10-31-simple-1.mdx

```diff
@@ -0,0 +1,102 @@
+
+On the other hand, writing vector functions has some drawbacks:
+
+- **Complexity** : Requires an understanding of Velox vectorized data representation and encodings, which requires additional work for our customers, specially those without DB background. And writing optimized vector functions requires even deeper understanding.
+- **Repetition** : Involves repeated efforts and code; in each function vectors are read in the same way, same set of optimizations are applied ... etc.
+- **Reliability** : More code means more bugs, especially in such a complex context.
+
+In the ideal world we would like to add most of the optimization that someone can do in a vector function to the simple functions adapter, so it would be enabled automatically. We have identified a number of optimizations that apply to all functions and implemented these generically in the SimpleFunctionAdapter. In this way, we can achieve the good of the two worlds and gain **Simplicity**, **Efficiency** and **Reliability** for most functions.
+
+In the past year we have been working on several improvements on the simple function interface on both the `expressivity` and `performance` axes that we will discuss in this series of notes. In this blogpost, we will talk about
+some of the general optimizations that we have in the adapter.
+
+## General optimizations
+
+#### Vector reuse
+If the output type matches one of the input types, and the input vector is to die after the function invocation, then it is possible to reuse it for the results instead of allocating a new vector.
+
+#### Bulk null setting
+Nulls are represented in a bit vector, hence, writing each bit can be expensive specially for primitive operations (like plus and minus). One optimization is
+to optimize for the `not null` case, and bulk setting the nulls to not null (64 bits in one instruction). After that during the computation, only if the results are null, the null bit is set to null.
+
+#### Null setting avoidance
+The adapter can statically infer if a function never generates null; In the simple function interface if the `call` function return's type is `void`, it means the output is never null, and if it's `bool`, then the function returns true for not null and false for null).
+When the function does not generate nulls, then null setting is **completely avoided** during the computation (only the previous bull setting is needed). The consequence of that is that the hot loop applying the function becomes simdizable triggering a huge boost in performance for primitive operations.
```

Comment:
I remember we had a number on how much faster this got, can we add this?

Reply:
s/bull/bulk

Reply:
There was also something related to the compiler being smartypants and constant folding the branch that sets null values if the function itself always returns a constant, providing essentially a similar perf boost. I think you were the one who pointed it out :)